astropy/coordinates/funcs.py

```diff
@@ -310,42 +312,31 @@ def concatenate_representations(reps):
     if not isinstance(reps, (Sequence, np.ndarray)):
         raise TypeError("Input must be a list or iterable of representation objects.")
 
-    # First, validate that the representations are the same, and
-    # concatenate all of the positional data:
     rep_type = type(reps[0])
-    if any(type(r) != rep_type for r in reps):
+    if any(type(rep) is not rep_type for rep in reps):
         raise TypeError("Input representations must all have the same type.")
 
-    # Construct the new representation with the concatenated data from the
-    # representations passed in
-    values = _concatenate_components(reps, rep_type.attr_classes.keys())
-    new_rep = rep_type(*values)
+    new_rep = _concatenate_components(reps, rep_type)
+
+    if all("s" not in rep.differentials for rep in reps):
+        return new_rep
 
-    has_diff = any("s" in rep.differentials for rep in reps)
-    if has_diff and any("s" not in rep.differentials for rep in reps):
+    try:
+        differentials = [rep.differentials["s"] for rep in reps]
+    except KeyError:
         raise ValueError(
             "Input representations must either all contain "
             "differentials, or not contain differentials."
-        )
-
-    if has_diff:
-        dif_type = type(reps[0].differentials["s"])
+        ) from None
 
-        if any(
-            "s" not in r.differentials or type(r.differentials["s"]) != dif_type
```

Comment:
The function already checked a few lines earlier that `"s"` is in all `r.differentials` and raised a `ValueError` if that wasn't the case, so checking it again here was redundant.